### ISSUE TITLE
Outbox table not created at runtime when name set in behavior

### ISSUE THREAD

**Issue by vince-nyanga** [Particular/NServiceBus.Persistence.AzureTable#717]
### Symptoms

The outbox table is not created and the following exception is logged:

`Azure.Data.Tables.TableTransactionFailedException : The specified resource does not exist.`

### Who's affected

Anyone setting the outbox table name in a [behavior](https://docs.particular.net/nservicebus/pipeline/manipulate-with-behaviors).

### Root cause

The code to create the table was missing.

<details>
I'm providing table information as part of the message handling pipeline, but the table is not created at runtime even though I have installers enabled. I can confirm that the table is created for sagas but not for outbox. 

I'm using `v5.0.0` and here's a snippet of my setup:
```csharp
endpointConfiguration.Pipeline.Register(new TableDetailsExtractionBehavior.RegistrationStep());

var tableServiceClient = new TableServiceClient(
            new Uri(tableStorageConfiguration.PrimaryEndpointUrl),
            new DefaultAzureCredential());
var persistence = endpointConfiguration.UsePersistence<AzureTablePersistence>();
persistence.UseTableServiceClient(tableServiceClient);
endpointConfiguration.EnableOutbox();
endpointConfiguration.EnableInstallers();
```
I don't know if I'm missing something, but the tables are not created as the [documentation](https://docs.particular.net/persistence/azure-table/configuration#table-name-configuration-and-creation-enabling-automatic-table-creation) suggests.
</details>

Backported to [5.0.2](https://github.com/Particular/NServiceBus.Persistence.AzureTable/pull/1029) and [6.0.1](https://github.com/Particular/NServiceBus.Persistence.AzureTable/pull/1028)

**Comment by vince-nyanga** [comment]
I'm also struggling to understand how the `SynchronizedStorageInstaller` will be run at runtime. I was under the impression that installers run at endpoint startup?

**Comment by abparticular** [comment]
@vince-nyanga I'm in the process of confirming whether there is an issue here.  Are you able to share the relevant parts of your `TableDetailsExtractionBehavior` and `RegistrationStep`?  It would be helpful to know where you are injecting your behavior into the pipeline.

**Comment by vince-nyanga** [comment]
@abparticular  here's an extract of `TableDetailsExtractionBehavior`
```csharp
public sealed class TableDetailsExtractionBehavior : Behavior<IIncomingLogicalMessageContext>
{
    public override Task Invoke(IIncomingLogicalMessageContext context, Func<Task> next)
    {
        var messageInstance = context.Message.Instance;
        if (TryGetTableAndPartitionkeyDetails(messageInstance, out var tableInformation, out var partitionKeyDetails))
        {
            context.Extensions.Set(tableInformation);
            context.Extensions.Set(partitionKeyDetails);
        }
        else
        {
            Log.Warning("Failed to extract table and partition key details from message {Message}", messageInstance.GetType().Name);
        }
        return next();
    }
//...
}

And the `RegistrationStep`:

```csharp
public sealed class RegistrationStep : RegisterStep
{
    public RegistrationStep()
        : base(nameof(TableDetailsExtractionBehavior),
            typeof(TableDetailsExtractionBehavior),
            "Extracts table information and partition key details from incoming message.",
            _ => new TableDetailsExtractionBehavior())
   {
   }
}
```
```

**Comment by abparticular** [comment]
@vince-nyanga 

Can you elaborate a bit on what's happening for you.  I understand you're saying that the table is not created at runtime, but are you getting any exceptions thrown?

> I can confirm that the table is created for sagas but not for outbox.

Since the saga and outbox data are stored in the same table, I am assuming you mean:
* when the outbox is not enabled then the saga table is created as you expect
* when the outbox is enabled, the table is no longer created.  
Is that correct?

If that is the case, do you have any other behaviors that you are registering in the pipeline?  The only thing I can think of is that the partition key is being supplied at the `ITransportReceiveContext` stage, without the associated table name also being supplied.  In this case a null reference exception will get thrown and the table won't be created (and we should look into throwing a more helpful exception).

To answer your other question, `SynchronizedStorageInstaller` is not run at runtime.  It does run at endpoint startup like you said.  The table gets created by the `AzureSagaPersister` at runtime if it detects it is necessary.

**Comment by vince-nyanga** [comment]
@abparticular , thanks for getting back to me.

> Can you elaborate a bit on what's happening for you. I understand you're saying that the table is not created at runtime, but are you getting any exceptions thrown?

An exception is thrown by the Azure Table Service indicating that the table I'm trying to write to does not exist.

> when the outbox is enabled, the table is no longer created.
Is that correct?

No, it's not correct. Let's say we have two messages -- `MessageA` which is handled by a saga, and `MessageB` which is not handled by a saga. If I dispatch `MessageA` the table is created at runtime by the `AzureSagaPersister`. However, if I dispatch `MessageB` the outbox persistence fails with an exception above. I could be wrong, but shouldn't we also have a similar method to `GetTableClientAndCreateTableIfNotExists()` in the `OutboxPersister` as well? Because the way I see it, we are assuming that all messages are handled by a saga hence we rely on the `AzureSagaPersister` to create the table if it doesn't exist. Unless if the `AzureSagaPersister` is also invoked in 'non-saga' messages?

**Comment by abparticular** [comment]
Thanks for the clarification @vince-nyanga.  I understand the issue now and am looking into how best to fix it.

**Comment by jpalac** [comment]
Will be shipped with Release 6.0.1 and 5.0.2

**Comment by jpalac** [comment]
5.0.2 and 6.0.1 have been [released ](https://discuss.particular.net/t/nservicebus-persistence-azuretable-6-0-1-and-5-0-2-patch-releases-available/4191)
